### ISSUE TITLE
etcd crash looping from panic

### ISSUE THREAD

**Issue by mattmattox** [rancher/rancher#19464]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
- Create 3 node Rancher v2.1.7 cluster using RKE using CentOS 7
- Do a rolling reboot of the cluster

**Result:**
etcd keeps crashing with the following error.
2019-04-04 18:00:31.867229 I | etcdserver: recovered store from snapshot at index 8100084
panic: invalid page allocation: 0
goroutine 1 [running]:
github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt.(*freelist).allocate(0xc42025d020, 0x51312d, 0x47, 0xc4203e4000)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt/freelist.go:85 +0x37b
github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt.(*DB).allocate(0xc420266000, 0x51312d, 0x47, 0xc4201f5920, 0x86043b, 0xc42025d020)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt/db.go:906 +0xa6
github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt.(*Tx).allocate(0xc42025a1c0, 0x47, 0x51312d, 0xc420251940, 0xc420251958)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt/tx.go:477 +0x4a
github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt.(*Tx).commitFreelist(0xc42025a1c0, 0x51312d, 0x7f2c9abed000)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt/tx.go:229 +0x8c
github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt.(*Tx).Commit(0xc42025a1c0, 0x33b1e007, 0x1516460)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/bbolt/tx.go:175 +0x676
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend.(*batchTx).commit(0xc42025d110, 0x0)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend/batch_tx.go:172 +0x82
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend.(*batchTxBuffered).unsafeCommit(0xc42025d110, 0xc4201f5b00)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend/batch_tx.go:245 +0x49
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend.(*batchTxBuffered).commit(0xc42025d110, 0xc4201f5b00)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend/batch_tx.go:232 +0x4c
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend.(*batchTxBuffered).Commit(0xc42025d110)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend/batch_tx.go:219 +0x40
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend.(*backend).ForceCommit(0xc42023b0a0)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/backend/backend.go:173 +0x2f
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc.NewStore(0x14ce160, 0xc42023b0a0, 0x14cefe0, 0x1536cd0, 0x14bac80, 0xc4203c12a0, 0xd3b9d)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/kvstore.go:127 +0x3ca
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc.newWatchableStore(0x14ce160, 0xc42023b0a0, 0x14cefe0, 0x1536cd0, 0x14bac80, 0xc4203c12a0, 0x1536cd0)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/watchable_store.go:75 +0x81
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc.New(0x14ce160, 0xc42023b0a0, 0x14cefe0, 0x1536cd0, 0x14bac80, 0xc4203c12a0, 0x1, 0xc4201f5e28)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/watchable_store.go:70 +0x5d
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver.recoverSnapshotBackend(0xc4201b6ea0, 0x14ce160, 0xc42023b0a0, 0xc4203d0500, 0x24b3, 0x2500, 0xc4203bf0a0, 0x3, 0x4, 0x0, ...)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/backend.go:74 +0xe0
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver.NewServer(0xc4201b6ea0, 0x0, 0x0, 0x0)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/server.go:378 +0x2dbe
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/embed.StartEtcd(0xc4201b8800, 0xc4201cd500, 0x0, 0x0)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/embed/etcd.go:157 +0x76a
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain.startEtcd(0xc4201b8800, 0x6, 0xf79a73, 0x6, 0x1)
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain/etcd.go:186 +0x58
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain.startEtcdOrProxyV2()
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain/etcd.go:103 +0x1579
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain.Main()
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain/main.go:39 +0x61
main.main()
/tmp/etcd-release-3.2.24/etcd/release/etcd/gopath/src/github.com/coreos/etcd/cmd/etcd/main.go:28 +0x20

**Other details that may be helpful:**
The workaround was to remove the node from the cluster using the following steps.
- Remove node from cluster.yml
- rke up
- run clean up script https://gist.github.com/superseb/2cf186726807a012af59a027cb41270d
- reboot node
- verify node is gone in etcdctl member list (run etcdctl member remove if needed)
- re-add node to cluster.yml
- rke up

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.1.7
- Installation option (single install/HA): HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Imported
- Machine type (cloud/VM/metal) and specifications (CPU/memory): 4/32GB

- Docker version (use `docker version`):

```
Containers: 22
 Running: 17
 Paused: 0
 Stopped: 5
Images: 10
Server Version: 18.09.3
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: e6b3f5632f50dbc4e9cb6288d911bf4f5e95b18e
runc version: 6635b4f0c6af3810594d2770f662f34ddc15b40d
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-957.5.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 31.26GiB
Name: ##########
ID: #######
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: bridge-nf-call-ip6tables is disabled
```

**Comment by alena1108** [comment]
@mattmattox can you elaborate on what `Do a rolling reboot of the cluster` means? How exactly do you trigger it

**Comment by mattmattox** [comment]
For the rolling rebooting, this environment was rebooted one server at a time using the steps listed below.
- Cordon node
- Drain node
- Sent reboot command
- Waited for node to show as healthy in UI
- Uncordon node
- Repeat for rest of cluster

**Comment by superseb** [comment]
Possible ref https://github.com/etcd-io/etcd/issues/10518

Seems like corruption happens and some data cannot be read anymore: https://github.com/etcd-io/bbolt/blob/v1.3.1-coreos.5/freelist.go#L85

Seems like etcd is also improving on freelist usage in boltdb/etcd as described in https://medium.com/@Alibaba_Cloud/performance-optimization-of-etcd-in-web-scale-data-scenario-b2eccae5d636.

Also there is no solid reproduce method at this time, as it seems to be corruption on a random place causing the panic to occur.

**Comment by alena1108** [comment]
@superseb thanks for the deep analysis. For now, we should monitor upstream etcd for the release containing all the needed fixes, and update Rancher to use it for k8s 1.15 (and 1.14, if available at that time)

**Comment by alena1108** [comment]
Per @superseb the fix will be most likely going to etcd 3.4

**Comment by deniseschannon** [comment]
etcd 3.4 was introduced in k8s 1.17.
